plugins/modules/nd_service_instance.py

```diff
@@ -0,0 +1,184 @@
+    instance_name: "default"
+    state: restart
+
+- name: Query a Service Instance with name and instance_name
+  cisco.nd.nd_service_instance:
+    name: "cisco-terraform"
+    instance_name: "default"
+    state: query
+  register: query_result
+
+- name: Query all Service Instances
+  cisco.nd.nd_service_instance:
+    state: query
+  register: query_results
+
+- name: Disable a Service Instance
+  cisco.nd.nd_service_instance:
+    name: "cisco-terraform"
+    instance_name: "default"
+    state: disable
+
+- name: Delete an existing service instance
+  cisco.nd.nd_service_instance:
+    name: "cisco-terraform"
+    target_version: "0.1.16"
+    state: delete
+"""
+
+RETURN = r"""
+"""
+
+from ansible.module_utils.basic import AnsibleModule
+from ansible_collections.cisco.nd.plugins.module_utils.nd import NDModule, nd_argument_spec
+
+
+def main():
+    argument_spec = nd_argument_spec()
+    argument_spec.update(
+        name=dict(type="str", aliases=["service_name"]),
+        instance_name=dict(type="str", aliases=["service_instance_name"]),
+        target_version=dict(type="str"),
+        state=dict(type="str", default="enable", choices=["enable", "query", "restart", "update", "disable", "delete"]),
+    )
+
+    module = AnsibleModule(
+        argument_spec=argument_spec,
+        supports_check_mode=True,
+        required_if=[
+            ["state", "enable", ["name", "instance_name", "target_version"]],
+            ["state", "update", ["name", "instance_name", "target_version"]],
+            ["state", "restart", ["name", "instance_name"]],
+            ["state", "disable", ["name", "instance_name"]],
+            ["state", "delete", ["name", "target_version"]],
+        ],
+    )
+
+    nd = NDModule(module)
+
+    name = nd.params.get("name")
+    target_version = nd.params.get("target_version")
+    instance_name = nd.params.get("instance_name")
+    state = nd.params.get("state")
+
+    base_path = "/nexus/infra/api/firmware/v1"
+    instance_path = "{0}/serviceinstances".format(base_path)
+
+    # Delete part
+    if state == "delete":
+        service_object = nd.query_obj("/nexus/infra/api/firmware/v1/services/{0}:{1}".format(name, target_version), ignore_not_found_error=True)
+        if service_object:
+            nd.existing = service_object
+    elif name and instance_name:
+        # Enable, Disable and Restart part
+        absent_instance_path = "{0}/serviceinstances/serviceName/{1}/instanceName/{2}".format(base_path, name, instance_name)
+        service_object = nd.query_obj(absent_instance_path, ignore_not_found_error=True)
+        if service_object:
+            nd.existing = service_object
+    else:
+        # Query all objects
+        nd.existing = nd.query_obj(instance_path)
+
+    nd.previous = nd.existing
+
+    unwanted = [
+        "metadata",
+        "status",
+        ["spec", "serviceReference"],
+    ]
+
+    if state == "enable" or (state == "update" and nd.existing):
```

Comment:
should we consider some error or warning handling for scenario's where we want to do an action but we don't execute code because the object doesn't exist in nd?

- state == "update" and not nd.existing
- state == "restart" and not nd.existing
- state == "disable" and not nd.existing